File: code-tiny/tools/flutter/normalizer.py
"""Map parser identities and facts into CortexHarness canonical graph rows."""

from __future__ import annotations

from dataclasses import dataclass, field
import hashlib
from typing import Any, Dict, Iterable, List, Mapping

from .models import AnalysisFacts, NodeRecord


CANONICAL_CLASS_KINDS = frozenset({"class", "mixin"})
CANONICAL_TYPE_KINDS = frozenset({"enum", "extension", "extension_type", "type_alias", "parameter"})
CANONICAL_FUNCTION_KINDS = frozenset(
    {"function", "method", "constructor", "getter", "setter", "operator"}
)


def stable_symbol_id(project_id: str, identity: str) -> str:
    """Return a project-scoped ID independent of the checkout location."""
    scope = project_id.strip()
    if not scope:
        raise ValueError("project_id must not be empty")
    digest = hashlib.sha256(f"dart\0{scope}\0{identity}".encode("utf-8")).hexdigest()[:24]
    return f"dart::{scope}::{digest}"


@dataclass
class CanonicalBatch:
    files: List[Dict[str, Any]] = field(default_factory=list)
    classes: List[Dict[str, Any]] = field(default_factory=list)
    types: List[Dict[str, Any]] = field(default_factory=list)
    functions: List[Dict[str, Any]] = field(default_factory=list)
    fields: List[Dict[str, Any]] = field(default_factory=list)
    relations: List[Dict[str, Any]] = field(default_factory=list)
    identity_to_id: Dict[str, str] = field(default_factory=dict)


def _base_row(
    node: NodeRecord,
    *,
    node_id: str,
    project_id: str,
    project_name: str,
    repo: str,
    build_system: str,
) -> Dict[str, Any]:
    properties = dict(node.properties)
    name = str(properties.get("name") or properties.get("path") or node.identity)
    qualified_name = str(properties.get("qualified_name") or name)
    code = str(properties.get("code") or "")
    comment = str(properties.get("comment") or "")
    return {
        "id": node_id,
        "name": name,
        "qualified_name": qualified_name,
        "kind": node.kind,
        "package_name": str(properties.get("package_uri") or properties.get("package_name") or ""),
        "class_name": str(properties.get("class_name") or ""),
        "scope_name": str(properties.get("scope_name") or properties.get("class_name") or ""),
        "file_path": node.evidence.file,
        "path": node.evidence.file,
        "start_byte": node.evidence.offset,
        "end_byte": node.evidence.offset + node.evidence.length,
        "start_line": node.evidence.start_line,
        "end_line": node.evidence.end_line,
        "arity": int(properties.get("arity", 0)),
        "code": code,
        "comment": comment,
        "summary": "",
        "note": str(properties.get("note") or comment or code[:4000]),
        "imports": list(properties.get("imports", [])),
        "exports": list(properties.get("exports", [])),
        "exported": bool(properties.get("exported", not name.startswith("_"))),
        "external": bool(properties.get("external", False)),
        "builtin": bool(properties.get("builtin", False)),
        "react_role": "",
        "middleware_kind": "",
        "type_signature": str(properties.get("type_signature") or ""),
        "project_id": project_id,
        "project_name": project_name,
        "language": "dart",
        "repo": repo,
        "build_system": build_system,
        "source_identity": node.identity,
        "generated": bool(properties.get("generated", False)),
    }


def normalize_facts(
    facts: AnalysisFacts,
    *,
    project_name: str | None = None,
    repo: str = "",
    build_system: str = "flutter",
    include_generated: bool = False,
) -> CanonicalBatch:
    project_id = facts.header.project_id
    batch = CanonicalBatch()
    included: List[NodeRecord] = []
    for node in facts.nodes:
        if bool(node.properties.get("generated", False)) and not include_generated:
            continue
        batch.identity_to_id[node.identity] = stable_symbol_id(project_id, node.identity)
        if bool(node.properties.get("_reference_only", False)):
            continue
        included.append(node)
    for node in included:
        row = _base_row(
            node,
            node_id=batch.identity_to_id[node.identity],
            project_id=project_id,
            project_name=project_name or project_id,
            repo=repo,
            build_system=build_system,
        )
        if node.kind == "file":
            batch.files.append(row)
        elif node.kind in CANONICAL_CLASS_KINDS:
            batch.classes.append(row)
        elif node.kind in CANONICAL_TYPE_KINDS:
            batch.types.append(row)
        elif node.kind in CANONICAL_FUNCTION_KINDS:
            batch.functions.append(row)
        elif node.kind == "field":
            batch.fields.append(row)
    for edge in facts.edges:
        source_id = batch.identity_to_id.get(edge.source)
        target_id = batch.identity_to_id.get(edge.target)
        if source_id is None or target_id is None:
            continue
        properties = dict(edge.properties)
        properties.update(
            {
                "confidence": edge.confidence,
                "source_file": edge.evidence.file,
                "source_line": edge.evidence.start_line,
                "project_id": project_id,
                "analyzer_version": facts.header.analyzer_version,
                "protocol_version": facts.header.schema_version,
            }
        )
        batch.relations.append(
            {
                "source_id": source_id,
                "target_id": target_id,
                "rel_type": edge.relationship,
                "properties": properties,
            }
        )
    for values in (batch.files, batch.classes, batch.types, batch.functions, batch.fields):
        values.sort(key=lambda row: row["id"])
    batch.relations.sort(key=lambda row: (row["rel_type"], row["source_id"], row["target_id"]))
    return batch


def qdrant_payloads(batch: CanonicalBatch) -> Iterable[Dict[str, Any]]:
    """Yield canonical symbol payloads for the existing project collection."""
    for row in (*batch.classes, *batch.types, *batch.functions, *batch.fields):
        yield {
            "symbol_id": row["id"],
            "qualified_name": row["qualified_name"],
            "name": row["name"],
            "kind": row["kind"],
            "file_path": row["file_path"],
            "start_line": row["start_line"],
            "end_line": row["end_line"],
            "code": row["code"],
            "comment": row["comment"],
            "note": row["note"],
            "project_id": row["project_id"],
            "project_name": row["project_name"],
            "language": "dart",
            "repo": row["repo"],
            "build_system": row["build_system"],
        }

File: rust/crates/analyzer-dart/src/normalizer.rs
//! Rust port của `tools/flutter/normalizer.py` — map parser identities và
//! facts thành canonical graph rows cho `LanguageCodeWriter::write_all`
//! (files/classes/types/functions/fields/relations).

use std::collections::{BTreeMap, BTreeSet};

use serde_json::{json, Map, Value};
use sha2::{Digest, Sha256};

use crate::models::{AnalysisFacts, NodeRecord, ANALYZER_VERSION, PROTOCOL_VERSION};

pub const CANONICAL_CLASS_KINDS: [&str; 2] = ["class", "mixin"];
pub const CANONICAL_TYPE_KINDS: [&str; 5] = ["enum", "extension", "extension_type", "type_alias", "parameter"];
pub const CANONICAL_FUNCTION_KINDS: [&str; 6] = [
    "function", "method", "constructor", "getter", "setter", "operator",
];

/// `stable_symbol_id` — project-scoped ID độc lập vị trí checkout.
pub fn stable_symbol_id(project_id: &str, identity: &str) -> Result<String, String> {
    let scope = project_id.trim();
    if scope.is_empty() {
        return Err("project_id must not be empty".to_string());
    }
    let mut hasher = Sha256::new();
    hasher.update(format!("dart\0{scope}\0{identity}"));
    let digest = hex_prefix(&hasher.finalize(), 24);
    Ok(format!("dart::{scope}::{digest}"))
}

fn hex_prefix(bytes: &[u8], len: usize) -> String {
    let mut out = String::with_capacity(len);
    for byte in bytes {
        if out.len() >= len {
            break;
        }
        out.push_str(&format!("{byte:02x}"));
    }
    out
}

/// `CanonicalBatch` — rows cho writer + mapping identity → symbol id.
#[derive(Default)]
pub struct CanonicalBatch {
    pub files: Vec<Map<String, Value>>,
    pub classes: Vec<Map<String, Value>>,
    pub types: Vec<Map<String, Value>>,
    pub functions: Vec<Map<String, Value>>,
    pub fields: Vec<Map<String, Value>>,
    pub relations: Vec<Map<String, Value>>,
    pub identity_to_id: BTreeMap<String, String>,
}

fn string_prop(properties: &BTreeMap<String, Value>, key: &str) -> String {
    properties
        .get(key)
        .and_then(Value::as_str)
        .unwrap_or_default()
        .to_string()
}

fn bool_prop(properties: &BTreeMap<String, Value>, key: &str, default: bool) -> bool {
    properties.get(key).and_then(Value::as_bool).unwrap_or(default)
}

fn string_list(properties: &BTreeMap<String, Value>, key: &str) -> Vec<Value> {
    properties
        .get(key)
        .and_then(Value::as_array)
        .cloned()
        .unwrap_or_default()
}

/// `_base_row` — 1:1 các key của Python normalizer.
#[allow(clippy::too_many_arguments)]
fn base_row(
    node: &NodeRecord,
    node_id: &str,
    project_id: &str,
    project_name: &str,
    repo: &str,
    build_system: &str,
) -> Map<String, Value> {
    let properties = &node.properties;
    let name = string_prop(properties, "name");
    let name = if name.is_empty() {
        string_prop(properties, "path")
    } else {
        name
    };
    let name = if name.is_empty() {
        node.identity.clone()
    } else {
        name
    };
    let qualified_name = {
        let value = string_prop(properties, "qualified_name");
        if value.is_empty() {
            name.clone()
        } else {
            value
        }
    };
    let code = string_prop(properties, "code");
    let comment = string_prop(properties, "comment");
    let note = {
        let value = string_prop(properties, "note");
        if !value.is_empty() {
            value
        } else if !comment.is_empty() {
            comment.clone()
        } else {
            code.chars().take(4000).collect()
        }
    };
    let package_name = {
        let value = string_prop(properties, "package_uri");
        if value.is_empty() {
            string_prop(properties, "package_name")
        } else {
            value
        }
    };
    let class_name = string_prop(properties, "class_name");
    let scope_name = {
        let value = string_prop(properties, "scope_name");
        if value.is_empty() {
            class_name.clone()
        } else {
            value
        }
    };
    let exported_default = !name.starts_with('_');
    let arity = properties.get("arity").and_then(Value::as_i64).unwrap_or(0);
    Map::from_iter([
        ("id".to_string(), json!(node_id)),
        ("name".to_string(), json!(name)),
        ("qualified_name".to_string(), json!(qualified_name)),
        ("kind".to_string(), json!(node.kind)),
        ("package_name".to_string(), json!(package_name)),
        ("class_name".to_string(), json!(class_name)),
        ("scope_name".to_string(), json!(scope_name)),
        ("file_path".to_string(), json!(node.evidence.file)),
        ("path".to_string(), json!(node.evidence.file)),
        ("start_byte".to_string(), json!(node.evidence.offset)),
        (
            "end_byte".to_string(),
            json!(node.evidence.offset + node.evidence.length),
        ),
        ("start_line".to_string(), json!(node.evidence.start_line)),
        ("end_line".to_string(), json!(node.evidence.end_line)),
        ("arity".to_string(), json!(arity)),
        ("code".to_string(), json!(code)),
        ("comment".to_string(), json!(comment)),
        ("summary".to_string(), json!("")),
        ("note".to_string(), json!(note)),
        ("imports".to_string(), Value::Array(string_list(properties, "imports"))),
        ("exports".to_string(), Value::Array(string_list(properties, "exports"))),
        (
            "exported".to_string(),
            json!(bool_prop(properties, "exported", exported_default)),
        ),
        ("external".to_string(), json!(bool_prop(properties, "external", false))),
        ("builtin".to_string(), json!(bool_prop(properties, "builtin", false))),
        ("react_role".to_string(), json!("")),
        ("middleware_kind".to_string(), json!("")),
        (
            "type_signature".to_string(),
            json!(string_prop(properties, "type_signature")),
        ),
        ("project_id".to_string(), json!(project_id)),
        ("project_name".to_string(), json!(project_name)),
        ("language".to_string(), json!("dart")),
        ("repo".to_string(), json!(repo)),
        ("build_system".to_string(), json!(build_system)),
        ("source_identity".to_string(), json!(node.identity)),
        ("generated".to_string(), json!(bool_prop(properties, "generated", false))),
    ])
}

/// `normalize_facts` — skip generated (trừ khi include_generated), giữ id
/// mapping cho endpoint `_reference_only`, bucket rows theo kind, sort id.
pub fn normalize_facts(
    facts: &AnalysisFacts,
    project_name: Option<&str>,
    repo: &str,
    build_system: &str,
) -> Result<CanonicalBatch, String> {
    let project_id = &facts.header.project_id;
    let mut batch = CanonicalBatch::default();
    let mut included: Vec<&NodeRecord> = Vec::new();
    for node in &facts.nodes {
        if bool_prop(&node.properties, "generated", false) {
            continue;
        }
        batch
            .identity_to_id
            .insert(node.identity.clone(), stable_symbol_id(project_id, &node.identity)?);
        if bool_prop(&node.properties, "_reference_only", false) {
            continue;
        }
        included.push(node);
    }
    let project_name = project_name.unwrap_or(project_id);
    for node in included {
        let row = base_row(
            node,
            &batch.identity_to_id[&node.identity],
            project_id,
            project_name,
            repo,
            build_system,
        );
        if node.kind == "file" {
            batch.files.push(row);
        } else if CANONICAL_CLASS_KINDS.contains(&node.kind.as_str()) {
            batch.classes.push(row);
        } else if CANONICAL_TYPE_KINDS.contains(&node.kind.as_str()) {
            batch.types.push(row);
        } else if CANONICAL_FUNCTION_KINDS.contains(&node.kind.as_str()) {
            batch.functions.push(row);
        } else if node.kind == "field" {
            batch.fields.push(row);
        }
    }
    for edge in &facts.edges {
        let Some(source_id) = batch.identity_to_id.get(&edge.source) else {
            continue;
        };
        let Some(target_id) = batch.identity_to_id.get(&edge.target) else {
            continue;
        };
        let mut properties = edge.properties.clone();
        properties.insert("confidence".to_string(), json!(edge.confidence));
        properties.insert("source_file".to_string(), json!(edge.evidence.file));
        properties.insert("source_line".to_string(), json!(edge.evidence.start_line));
        properties.insert("project_id".to_string(), json!(project_id));
        properties.insert("analyzer_version".to_string(), json!(ANALYZER_VERSION));
        properties.insert("protocol_version".to_string(), json!(PROTOCOL_VERSION));
        let properties_map: Map<String, Value> = properties.into_iter().collect();
        batch.relations.push(Map::from_iter([
            ("source_id".to_string(), json!(source_id)),
            ("target_id".to_string(), json!(target_id)),
            ("rel_type".to_string(), json!(edge.relationship)),
            ("properties".to_string(), Value::Object(properties_map)),
        ]));
    }
    for values in [
        &mut batch.files,
        &mut batch.classes,
        &mut batch.types,
        &mut batch.functions,
        &mut batch.fields,
    ] {
        values.sort_by(|left, right| {
            left.get("id")
                .and_then(Value::as_str)
                .unwrap_or_default()
                .cmp(right.get("id").and_then(Value::as_str).unwrap_or_default())
        });
    }
    batch.relations.sort_by(|left, right| {
        let key = |row: &Map<String, Value>| {
            (
                row.get("rel_type").and_then(Value::as_str).unwrap_or_default().to_string(),
                row.get("source_id").and_then(Value::as_str).unwrap_or_default().to_string(),
                row.get("target_id").and_then(Value::as_str).unwrap_or_default().to_string(),
            )
        };
        key(left).cmp(&key(right))
    });
    Ok(batch)
}

/// Gom id của mọi symbol row (keep_ids cho incremental cleanup flutter).
pub fn symbol_keep_ids(batch: &CanonicalBatch) -> BTreeSet<String> {
    [&batch.files, &batch.classes, &batch.types, &batch.functions, &batch.fields]
        .into_iter()
        .flatten()
        .filter_map(|row| row.get("id").and_then(Value::as_str).map(str::to_string))
        .collect()
}
